In [2]:
import pandas as pd

# Exercise 1

In [10]:
customer_df = pd.read_csv("adventure works 2020_customer.csv")

In [11]:
customer_df['BirthDate'] = pd.to_datetime(customer_df['BirthDate'])

In [61]:
#Calculation of customers age
customer_df['CustomerAge'] = (pd.to_datetime("today").year) - (customer_df["BirthDate"].dt.year)
customer_df['CustomerAge']

0        39
1        40
2        40
3        37
4        37
         ..
18479    47
18480    45
18481    45
18482    46
18483    46
Name: CustomerAge, Length: 18484, dtype: int32

In [62]:
#Split customers age into 4 span
bins = [1, 25, 40, 60, 100]
labels = [' < 25', '25 - 40', '40 - 60', ' >= 60']
customer_df['AgeSpan'] = pd.cut(customer_df['CustomerAge'], bins= bins, labels= labels, right = False)

In [64]:
#Calculation of number of customers in 4 span age
customer_df.groupby('AgeSpan', as_index= False).agg(number = ('CustomerAge', 'count'))

C:\Users\ASUS\AppData\Local\Temp\ipykernel_220236\576280510.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  customer_df.groupby('AgeSpan', as_index= False).agg(number = ('CustomerAge', 'count'))


AgeSpan  number
0     < 25       0
1  25 - 40    7702
2  40 - 60    8960
3    >= 60    1822

There is no customer under age 25. **90%** customers are between **25 to 60** years old and gender of them are half male and half female.

# Exeercise 2

In [3]:
sales_df = pd.read_csv('Adventure works 2020_sales.csv')

In [4]:
sales_df['OrderDate'] = pd.to_datetime(sales_df['OrderDate'])

In [67]:
#Finding weekday of every sale
sales_df['WeekDay'] = sales_df['OrderDate'].dt.day_name()
sales_df['WeekDay'].unique()

array(['Friday', 'Saturday', 'Sunday', 'Monday', 'Tuesday', 'Wednesday',
       'Thursday'], dtype=object)

In [68]:
#Calculation of sales base on weekdays
sales_df.groupby('WeekDay').agg(number = ('WeekDay', 'count')).sort_values(by = 'number', ascending= False)

number
WeekDay          
Wednesday    8971
Friday       8755
Thursday     8754
Monday       8611
Tuesday      8529
Sunday       8518
Saturday     8260

Dispersion of number of sales is nearly close to each other. Most sales, 15%, are on the day **wednesday** during 2016 to 2019.

# Exercise 3

In [69]:
product_df = pd.read_csv('adventure works 2020_product.csv')

C:\Users\ASUS\AppData\Local\Temp\ipykernel_220236\451194970.py:1: DtypeWarning: Columns (2,4,8,11,13,14,15,16,17,18,19,20) have mixed types. Specify dtype option on import or set low_memory=False.
  product_df = pd.read_csv('adventure works 2020_product.csv')


In [70]:
#Mask sales on 2018
sales_on_2018 = sales_df[sales_df['OrderDate'].dt.year == 2018]
sales_on_2018.head()

OrderDate  OrderDate Key  ProductKey  CustomerKey  SalesTerritoryKey  \
3690 2018-01-01       20180101         352        26708                  4   
3691 2018-01-01       20180101         385        14335                  1   
3692 2018-01-01       20180101         327        15493                  1   
3693 2018-01-01       20180101         375        16538                  8   
3694 2018-01-01       20180101         375        15094                  7   

     SalesOrderNumber  ShipDate  SalesOrderLineNumber  OrderQuantity  \
3690          SO48799  1/8/2018                     1              1   
3691          SO48797  1/8/2018                     1              1   
3692          SO48801  1/8/2018                     1              1   
3693          SO48795  1/8/2018                     1              1   
3694          SO48796  1/8/2018                     1              1   

      UnitPrice  ExtendedAmount  UnitPriceDiscountPct  DiscountAmount  \
3690  2071.4196       2071.4196                     0               0   
3691  1000.4375       1000.4375                     0               0   
3692   782.9900        782.9900                     0               0   
3693  2181.5625       2181.5625                     0               0   
3694  2181.5625       2181.5625                     0               0   

      ProductStandardCost  TotalProductCost  SalesAmount    TaxAmt  Freight  \
3690            1117.8559         1117.8559    2071.4196  165.7136  51.7855   
3691             605.6492          605.6492    1000.4375   80.0350  25.0109   
3692             486.7066          486.7066     782.9900   62.6392  19.5748   
3693            1320.6838         1320.6838    2181.5625  174.5250  54.5391   
3694            1320.6838         1320.6838    2181.5625  174.5250  54.5391   

     RegionMonthID WeekDay  
3690    Southwest1  Monday  
3691    Northwest1  Monday  
3692    Northwest1  Monday  
3693      Germany1  Monday  
3694       France1  Monday

In [71]:
sale_merge_with_product = sales_on_2018.merge(product_df[['ProductKey', 'ProductName']], on = 'ProductKey', how= 'left')
sale_merge_with_product.head()

OrderDate  OrderDate Key  ProductKey  CustomerKey  SalesTerritoryKey  \
0 2018-01-01       20180101         352        26708                  4   
1 2018-01-01       20180101         385        14335                  1   
2 2018-01-01       20180101         327        15493                  1   
3 2018-01-01       20180101         375        16538                  8   
4 2018-01-01       20180101         375        15094                  7   

  SalesOrderNumber  ShipDate  SalesOrderLineNumber  OrderQuantity  UnitPrice  \
0          SO48799  1/8/2018                     1              1  2071.4196   
1          SO48797  1/8/2018                     1              1  1000.4375   
2          SO48801  1/8/2018                     1              1   782.9900   
3          SO48795  1/8/2018                     1              1  2181.5625   
4          SO48796  1/8/2018                     1              1  2181.5625   

   ...  UnitPriceDiscountPct  DiscountAmount  ProductStandardCost  \
0  ...                     0               0            1117.8559   
1  ...                     0               0             605.6492   
2  ...                     0               0             486.7066   
3  ...                     0               0            1320.6838   
4  ...                     0               0            1320.6838   

   TotalProductCost  SalesAmount    TaxAmt  Freight  RegionMonthID WeekDay  \
0         1117.8559    2071.4196  165.7136  51.7855     Southwest1  Monday   
1          605.6492    1000.4375   80.0350  25.0109     Northwest1  Monday   
2          486.7066     782.9900   62.6392  19.5748     Northwest1  Monday   
3         1320.6838    2181.5625  174.5250  54.5391       Germany1  Monday   
4         1320.6838    2181.5625  174.5250  54.5391        France1  Monday   

               ProductName  
0  Mountain-200 Silver, 38  
1    Road-550-W Yellow, 42  
2         Road-650 Red, 44  
3       Road-250 Black, 48  
4       Road-250 Black, 48  

[5 rows x 21 columns]

In [72]:
#Counting sales base on each product category
product_sales = sale_merge_with_product.groupby('ProductName')['OrderQuantity'].agg(order_quantity = 'sum').sort_values(by = 'order_quantity' , ascending= False)
product_sales.head(10)

order_quantity
ProductName                            
Water Bottle - 30 oz.              1742
Patch Kit/8 Patches                1356
Mountain Tire Tube                 1313
Road Tire Tube                      999
Sport-100 Helmet, Red               898
Sport-100 Helmet, Black             891
AWC Logo Cap                        885
Fender Set - Mountain               883
Sport-100 Helmet, Blue              857
Mountain Bottle Cage                824

**Water bottle** is in the first rank of order quantity. Products related to Bike are on the other ranks.

# Exercise 4

In [73]:
product_subcategory_df = pd.read_csv('adventure works 2020_dimproductsubcategory.csv')
product_category_df = pd.read_csv('adventure works 2020_dimproductcategory .csv')

In [74]:
product_category_df.head()

ProductCategoryKey  ProductCategoryAlternateKey EnglishProductCategoryName  \
0                   1                            1                      Bikes   
1                   2                            2                 Components   
2                   3                            3                   Clothing   
3                   4                            4                Accessories   

  SpanishProductCategoryName FrenchProductCategoryName  
0                  Bicicleta                      Vélo  
1                 Componente                 Composant  
2                     Prenda                 Vêtements  
3                  Accesorio                Accessoire

In [75]:
product_subcategory_df.head()

ProductSubcategoryKey  ProductSubcategoryAlternateKey  \
0                      1                               1   
1                      2                               2   
2                      3                               3   
3                      4                               4   
4                      5                               5   

  EnglishProductSubcategoryName SpanishProductSubcategoryName  \
0                Mountain Bikes          Bicicleta de montaña   
1                    Road Bikes        Bicicleta de carretera   
2                 Touring Bikes            Bicicleta de paseo   
3                    Handlebars                         Barra   
4               Bottom Brackets               Eje de pedalier   

  FrenchProductSubcategoryName  ProductCategoryKey  
0                          VTT                   1  
1                Vélo de route                   1  
2            Vélo de randonnée                   1  
3                Barre d'appui                   2  
4              Axe de pédalier                   2

In [76]:
product_merge_with_subcategory = product_df.merge(product_subcategory_df, on = 'ProductSubcategoryKey', how = 'left')
product_merge_with_subcategory.head()

ProductKey  ProductSubcategoryKey                ProductName  StandardCost  \
0       362.0                    1.0     Mountain-200 Black, 46     1105.8100   
1       363.0                    1.0     Mountain-200 Black, 46     1251.9813   
2       364.0                    1.0     Mountain-300 Black, 38      598.4354   
3       365.0                    1.0     Mountain-300 Black, 40      598.4354   
4       587.0                    1.0  Mountain-400-W Silver, 38      419.7784   

    Color  SafetyStockLevel  ListPrice  Size SizeRange  Weight  ...   Status  \
0   Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...      NaN   
1   Black             100.0  2294.9900  46.0  42-46 CM   24.13  ...  Current   
2   Black             100.0  1079.9900  38.0  38-40 CM   25.35  ...      NaN   
3   Black             100.0  1079.9900  40.0  38-40 CM   25.77  ...      NaN   
4  Silver             100.0   769.4900  38.0  38-40 CM   26.35  ...  Current   

      SubCategory  Category StartDate    EndDate  \
0  Mountain Bikes     Bikes  7/1/2017  6/30/2018   
1  Mountain Bikes     Bikes  7/1/2018        NaN   
2  Mountain Bikes     Bikes  7/1/2017  6/30/2018   
3  Mountain Bikes     Bikes  7/1/2017  6/30/2018   
4  Mountain Bikes     Bikes  7/1/2018        NaN   

  ProductSubcategoryAlternateKey EnglishProductSubcategoryName  \
0                            1.0                Mountain Bikes   
1                            1.0                Mountain Bikes   
2                            1.0                Mountain Bikes   
3                            1.0                Mountain Bikes   
4                            1.0                Mountain Bikes   

  SpanishProductSubcategoryName FrenchProductSubcategoryName  \
0          Bicicleta de montaña                          VTT   
1          Bicicleta de montaña                          VTT   
2          Bicicleta de montaña                          VTT   
3          Bicicleta de montaña                          VTT   
4          Bicicleta de montaña                          VTT   

  ProductCategoryKey  
0                1.0  
1                1.0  
2                1.0  
3                1.0  
4                1.0  

[5 rows x 26 columns]

In [77]:
product_subcategory_sale_merge = product_merge_with_subcategory.merge(sales_df, on = 'ProductKey', how = 'left')
product_subcategory_sale_merge.head()

ProductKey  ProductSubcategoryKey             ProductName  StandardCost  \
0       362.0                    1.0  Mountain-200 Black, 46       1105.81   
1       362.0                    1.0  Mountain-200 Black, 46       1105.81   
2       362.0                    1.0  Mountain-200 Black, 46       1105.81   
3       362.0                    1.0  Mountain-200 Black, 46       1105.81   
4       362.0                    1.0  Mountain-200 Black, 46       1105.81   

   Color  SafetyStockLevel  ListPrice  Size SizeRange  Weight  ...  \
0  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
1  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
2  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
3  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
4  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   

   ExtendedAmount UnitPriceDiscountPct  DiscountAmount ProductStandardCost  \
0       2049.0982                  0.0             0.0             1105.81   
1       2049.0982                  0.0             0.0             1105.81   
2       2049.0982                  0.0             0.0             1105.81   
3       2049.0982                  0.0             0.0             1105.81   
4       2049.0982                  0.0             0.0             1105.81   

  TotalProductCost SalesAmount    TaxAmt  Freight    RegionMonthID   WeekDay  
0          1105.81   2049.0982  163.9279  51.2275       Australia7  Saturday  
1          1105.81   2049.0982  163.9279  51.2275  United Kingdom7  Thursday  
2          1105.81   2049.0982  163.9279  51.2275       Australia7    Friday  
3          1105.81   2049.0982  163.9279  51.2275       Australia7   Tuesday  
4          1105.81   2049.0982  163.9279  51.2275       Southwest7    Friday  

[5 rows x 45 columns]

In [78]:
product_merge_with_category = product_merge_with_subcategory.merge(product_category_df, on = 'ProductCategoryKey', how = 'left')
product_merge_with_category.head()

ProductKey  ProductSubcategoryKey                ProductName  StandardCost  \
0       362.0                    1.0     Mountain-200 Black, 46     1105.8100   
1       363.0                    1.0     Mountain-200 Black, 46     1251.9813   
2       364.0                    1.0     Mountain-300 Black, 38      598.4354   
3       365.0                    1.0     Mountain-300 Black, 40      598.4354   
4       587.0                    1.0  Mountain-400-W Silver, 38      419.7784   

    Color  SafetyStockLevel  ListPrice  Size SizeRange  Weight  ...  \
0   Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
1   Black             100.0  2294.9900  46.0  42-46 CM   24.13  ...   
2   Black             100.0  1079.9900  38.0  38-40 CM   25.35  ...   
3   Black             100.0  1079.9900  40.0  38-40 CM   25.77  ...   
4  Silver             100.0   769.4900  38.0  38-40 CM   26.35  ...   

     EndDate ProductSubcategoryAlternateKey  EnglishProductSubcategoryName  \
0  6/30/2018                            1.0                 Mountain Bikes   
1        NaN                            1.0                 Mountain Bikes   
2  6/30/2018                            1.0                 Mountain Bikes   
3  6/30/2018                            1.0                 Mountain Bikes   
4        NaN                            1.0                 Mountain Bikes   

  SpanishProductSubcategoryName FrenchProductSubcategoryName  \
0          Bicicleta de montaña                          VTT   
1          Bicicleta de montaña                          VTT   
2          Bicicleta de montaña                          VTT   
3          Bicicleta de montaña                          VTT   
4          Bicicleta de montaña                          VTT   

  ProductCategoryKey ProductCategoryAlternateKey EnglishProductCategoryName  \
0                1.0                         1.0                      Bikes   
1                1.0                         1.0                      Bikes   
2                1.0                         1.0                      Bikes   
3                1.0                         1.0                      Bikes   
4                1.0                         1.0                      Bikes   

  SpanishProductCategoryName FrenchProductCategoryName  
0                  Bicicleta                      Vélo  
1                  Bicicleta                      Vélo  
2                  Bicicleta                      Vélo  
3                  Bicicleta                      Vélo  
4                  Bicicleta                      Vélo  

[5 rows x 30 columns]

In [79]:
product_category_sale_merge = product_merge_with_category.merge(sales_df, on = 'ProductKey', how = 'left')
product_category_sale_merge.head()

ProductKey  ProductSubcategoryKey             ProductName  StandardCost  \
0       362.0                    1.0  Mountain-200 Black, 46       1105.81   
1       362.0                    1.0  Mountain-200 Black, 46       1105.81   
2       362.0                    1.0  Mountain-200 Black, 46       1105.81   
3       362.0                    1.0  Mountain-200 Black, 46       1105.81   
4       362.0                    1.0  Mountain-200 Black, 46       1105.81   

   Color  SafetyStockLevel  ListPrice  Size SizeRange  Weight  ...  \
0  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
1  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
2  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
3  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   
4  Black             100.0  2049.0982  46.0  42-46 CM   24.13  ...   

   ExtendedAmount UnitPriceDiscountPct  DiscountAmount ProductStandardCost  \
0       2049.0982                  0.0             0.0             1105.81   
1       2049.0982                  0.0             0.0             1105.81   
2       2049.0982                  0.0             0.0             1105.81   
3       2049.0982                  0.0             0.0             1105.81   
4       2049.0982                  0.0             0.0             1105.81   

  TotalProductCost SalesAmount    TaxAmt  Freight    RegionMonthID   WeekDay  
0          1105.81   2049.0982  163.9279  51.2275       Australia7  Saturday  
1          1105.81   2049.0982  163.9279  51.2275  United Kingdom7  Thursday  
2          1105.81   2049.0982  163.9279  51.2275       Australia7    Friday  
3          1105.81   2049.0982  163.9279  51.2275       Australia7   Tuesday  
4          1105.81   2049.0982  163.9279  51.2275       Southwest7    Friday  

[5 rows x 49 columns]

In [80]:
#Calculation of average sale base on every product subcategory
mean_price_subcategory = product_subcategory_sale_merge.groupby('EnglishProductSubcategoryName')['UnitPrice'].agg(average_price = 'mean').sort_values(by = 'average_price', ascending= False)
mean_price_subcategory.head()

average_price
EnglishProductSubcategoryName               
Mountain Bikes                   2002.567317
Road Bikes                       1799.774918
Touring Bikes                    1774.250600
Bike Stands                       159.000000
Bike Racks                        120.000000

In [81]:
#Calculation of average sale base on every product category
mean_price_category = product_category_sale_merge.groupby('EnglishProductCategoryName')['UnitPrice'].agg(average_price = 'mean').sort_values(by = 'average_price',ascending= False)
mean_price_category.head()

average_price
EnglishProductCategoryName               
Bikes                         1862.423193
Clothing                        37.333547
Accessories                     19.415936
Components                            NaN

There are 4 category for each released product that each category conclude subcategories. Base on category, ***bike*** products are more interested for customers. So ***subcategory of bike*** are also more favorable regard to other products.

# Exercise 5 (method 1)

In [99]:
sales_df['OrderDate'] = pd.to_datetime(sales_df['OrderDate'])

In [5]:
#Identify last order of each customer
max_order_date = sales_df['OrderDate'].max()
max_order_date

Timestamp('2019-07-31 00:00:00')

In [8]:
six_months_ago = max_order_date - pd.DateOffset(months= 6)
six_months_ago

Timestamp('2019-01-31 00:00:00')

In [9]:
#Identify each customer doesnt purchase over past 6 month (6 month is calculated base on last date of order date (2019-07-31))
active_customers = sales_df[sales_df['OrderDate'] > six_months_ago]['CustomerKey'].unique()
active_customers

array([14748, 16017, 16677, ..., 21524, 18178, 18529], shape=(9953,))

In [100]:
inactive_customers = customer_df[~customer_df['CustomerKey'].isin(active_customers)]
inactive_customers.head()

CustomerKey  GeographyKey               Name  BirthDate MaritalStatus  \
0        11000            26           Jon Yang 1986-04-08             M   
2        11002            31       Ruben Torres 1985-08-12             M   
3        11003            11        Christy Zhu 1988-02-15             S   
4        11004            19  Elizabeth Johnson 1988-08-08             S   
5        11005            22         Julio Ruiz 1985-08-05             S   

  Gender  YearlyIncome  NumberChildrenAtHome    Occupation  HouseOwnerFlag  \
0      M         90000                     0  Professional               1   
2      M         60000                     3  Professional               1   
3      F         70000                     0  Professional               0   
4      F         80000                     5  Professional               1   
5      M         70000                     0  Professional               1   

   NumberCarsOwned         AddressLine1 AddressLine2                Phone  \
0                0      3761 N. 14th St          NaN  1 (11) 500 555-0162   
2                1     5844 Linden Land          NaN  1 (11) 500 555-0184   
3                1     1825 Village Pl.          NaN  1 (11) 500 555-0162   
4                4  7553 Harness Circle          NaN  1 (11) 500 555-0131   
5                1  7305 Humphrey Drive          NaN  1 (11) 500 555-0151   

  DateFirstPurchase  
0    7/22/2016 0:00  
2    7/10/2016 0:00  
3     7/1/2016 0:00  
4    7/26/2016 0:00  
5     7/2/2016 0:00

In [101]:
#Identification of inactive customers over past 6 months
six_month_inactive = inactive_customers[['CustomerKey', 'Name']].reset_index().drop(columns= 'index')
six_month_inactive.tail()


CustomerKey           Name
8526        29479     Tommy Tang
8527        29480      Nina Raji
8528        29481      Ivan Suri
8529        29482  Clayton Zhang
8530        29483  Jésus Navarro

# Exercise 5 (method 2)

In [109]:
#Identify last order of each customer
last_order = sales_df.groupby('CustomerKey')['OrderDate'].agg(last_order = 'max').reset_index()
last_order.head()

CustomerKey last_order
0        11000 2018-11-04
1        11001 2019-06-12
2        11002 2018-08-27
3        11003 2018-11-11
4        11004 2018-11-02

In [115]:
#Identify each customer doesnt purchase over past 6 month (6 month is calculated base on last date of order date (2019-07-31))
inactive_customers = last_order[((sales_df['OrderDate'].max() - last_order['last_order']).dt.days) > 180]
inactive_customers.head()

CustomerKey last_order
0        11000 2018-11-04
2        11002 2018-08-27
3        11003 2018-11-11
4        11004 2018-11-02
5        11005 2018-11-03

In [111]:
customer_merge_with_inactives = inactive_customers.merge(customer_df, on = 'CustomerKey', how = 'left')
customer_merge_with_inactives.tail()

CustomerKey last_order  GeographyKey           Name  BirthDate  \
8526        29479 2018-03-08           209     Tommy Tang 1978-07-04   
8527        29480 2019-01-18           248      Nina Raji 1980-11-10   
8528        29481 2017-02-13           120      Ivan Suri 1980-01-05   
8529        29482 2018-03-22           179  Clayton Zhang 1979-03-05   
8530        29483 2018-03-13           217  Jésus Navarro 1979-12-08   

     MaritalStatus Gender  YearlyIncome  NumberChildrenAtHome Occupation  \
8526             M      M         30000                     0   Clerical   
8527             S      F         30000                     0   Clerical   
8528             S      M         30000                     0   Clerical   
8529             M      M         30000                     0   Clerical   
8530             M      M         30000                     0   Clerical   

      HouseOwnerFlag  NumberCarsOwned               AddressLine1 AddressLine2  \
8526               1                0          111, rue Maillard          NaN   
8527               1                0          9 Katherine Drive          NaN   
8528               0                0                Knaackstr 4          NaN   
8529               1                0     1080, quai de Grenelle          NaN   
8530               1                0  244, rue de la Centenaire          NaN   

                    Phone DateFirstPurchase  
8526  1 (11) 500 555-0136     3/8/2018 0:00  
8527  1 (11) 500 555-0146    1/18/2019 0:00  
8528  1 (11) 500 555-0144    2/13/2017 0:00  
8529  1 (11) 500 555-0137    3/22/2018 0:00  
8530  1 (11) 500 555-0141    3/13/2018 0:00

In [113]:
six_month_inactive = customer_merge_with_inactives[['CustomerKey', 'Name']].reset_index().drop(columns = 'index')
six_month_inactive.tail()

CustomerKey           Name
8526        29479     Tommy Tang
8527        29480      Nina Raji
8528        29481      Ivan Suri
8529        29482  Clayton Zhang
8530        29483  Jésus Navarro

An analysis of **inactive customers** over the past six months revealed that out of total of 18,484 customers, 8,530 individuals ,equivalent to **46%**, had not made any purchase during this period. Among these inactive customers, approximately half were male and the other half were female. 